src/librustc_typeck/check/upvar.rs

```diff
@@ -407,15 +412,19 @@ impl<'a, 'gcx, 'tcx> AdjustBorrowKind<'a, 'gcx, 'tcx> {
                 }
 
                 // also need to be in an FnMut closure since this is not an ImmBorrow
-                self.adjust_closure_kind(upvar_id.closure_expr_id, ty::ClosureKind::FnMut);
+                self.adjust_closure_kind(upvar_id.closure_expr_id,
+                                         ty::ClosureKind::FnMut,
+                                         tcx.hir.span(upvar_id.var_id));
 
                 true
             }
             mc::NoteClosureEnv(upvar_id) => {
                 // this kind of deref occurs in a `move` closure, or
                 // for a by-value upvar; in either case, to mutate an
                 // upvar, we need to be an FnMut closure
-                self.adjust_closure_kind(upvar_id.closure_expr_id, ty::ClosureKind::FnMut);
+                self.adjust_closure_kind(upvar_id.closure_expr_id,
+                                         ty::ClosureKind::FnMut,
+                                         tcx.hir.span(upvar_id.var_id));
```

Comment:
here, in addition to passing in the `&Note`, we can pass in the `cmt` that the note originated from, and use that span. That should be the span of the variable *reference*.